molecule/create.yml

```diff
@@ -0,0 +1,166 @@
+      loop: "{{ molecule_yml.platforms | selectattr('name', 'eq', 'ns') | list }}"
+      async: 7200
+      poll: 0
+
+    - name: Wait for DNS creation to complete
+      async_status:
+        jid: "{{ item.ansible_job_id }}"
+      register: dnsdocker_jobs
+      until: dnsdocker_jobs.finished
+      retries: 300
+      loop: "{{ dnsserver.results }}"
+
+    - name: Add DNS container to inventory
+      add_host:
+        name: ns
+        ansible_connection: docker
+      changed_when: false
+
+    - name: Recover internal_dns_ip (from ns container)
+      set_fact:
+        internal_dns_ip: "{{ dnsdocker_jobs.results[0].ansible_facts.docker_container.NetworkSettings.Networks.external.IPAddress }}"
+
+    - name: Create molecule instance(s)
+      docker_container:
+        name: "{{ item.name }}"
+        docker_host: "{{ item.docker_host | default(lookup('env', 'DOCKER_HOST') or 'unix://var/run/docker.sock') }}"
+        hostname: "{{ item.hostname | default(item.name) }}"
+        image: "{{ item.pre_build_image | default(false) | ternary('', 'molecule_local/') }}{{ item.image }}"
+        state: started
+        recreate: false
+        log_driver: json-file
+        command: "{{ item.command | default('bash -c \"while true; do sleep 10000; done\"') }}"
+        privileged: "{{ item.privileged | default(omit) }}"
+        security_opts: "{{ item.security_opts | default(omit) }}"
+        volumes: "{{ item.volumes | default(omit) }}"
+        tmpfs: "{{ item.tmpfs | default(omit) }}"
+        capabilities: "{{ item.capabilities | default(omit) }}"
+        exposed_ports: "{{ item.exposed_ports | default(omit) }}"
+        published_ports: "{{ item.published_ports | default(omit) }}"
+        ulimits: "{{ item.ulimits | default(omit) }}"
+        networks: "{{ item.networks | default(omit) }}"
+        network_mode: "{{ item.network_mode | default(omit) }}"
+        dns_servers: "{{ internal_dns_ip }}"
+        env: "{{ item.env | default(omit) }}"
+        restart_policy: "{{ item.restart_policy | default(omit) }}"
+        restart_retries: "{{ item.restart_retries | default(omit) }}"
+      register: server
+      loop: "{{ molecule_yml.platforms | selectattr('name', 'ne', 'ns') | list }}"
+      async: 7200
+      poll: 0
+
+    - name: Wait for instance(s) creation to complete
+      async_status:
+        jid: "{{ item.ansible_job_id }}"
+      register: docker_jobs
+      until: docker_jobs.finished
+      retries: 300
+      loop: "{{ server.results }}"
+
+    - name: Print docker_jobs (debug only)
+      debug:
+        var: docker_jobs
+        verbosity: 2
+
+    - name: Add mail.not.local DNS entry
+      lineinfile:
+        line: "{{ item.ansible_facts.docker_container.NetworkSettings.Networks.external.Gateway }}   mail.not.local"
+        state: present
+        path: /var/dns-hosts/molecule
+        create: true
+      loop: "{{docker_jobs.results}}"
+      delegate_to: ns
+
+    - name: Add per-container DNS entry
```

Comment:
I wasn't convinced when you first explained it to me, but I've got to admit, this way out of the cyclic dependency between DNS server IP, and other container's IPs is pretty elegant.

However there's one thing I don't understand : how does the config get reloaded ? Does dnsmasq constantly watch its configuration file for changes ?

Reply:
Yes, dnsmasq does listen for host files and does not need reloading. This should be documented with a comment.

Reply:
This behaviour is now described in a comment